### PR TITLE
Update plugin(s) build setup for distribution

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary

Fixes task 3 of https://github.com/WordPress/performance/issues/1012

As mentioned by @westonruter in https://github.com/WordPress/performance/issues/893#issuecomment-1947398713, we are now moving towards use cases where we can have to build files that are not in the git tree and hence will be ignored by `git archive`.

Also as outlined in https://github.com/WordPress/performance/pull/1024#issuecomment-1979221706, in the new mono repo setup we will require a build setup because the repo level `.gitattributes` or `.distignore` won't work just to ignore files from standalone distributed plugins.

This PR proposes a simple Webpack-based plugin bundling solution that solves both problems.
## Relevant technical choices

- Switch to Webpack for handling plugin(s) building tasks with `copy-webpack-plugin` Webpack plugin.
- Remove dependency from custom-written CLI scripts to:
    - Get the plugin version from the `readme.txt` file. This task is now handled by Webpack which creates a `manifest.json` with plugin versions.
    - Get the plugin path. This is not required as we are completely removing the modules.
    - Build the standalone plugins. This task is now handled by Webpack.
- Add plugin(s) deployment status on the repo.
<details><summary>Screenshot</summary>

image
</details> 

- Add a workflow step to add the plugin(s) built to release assets with their checksum to ensure assets' integrity.

<details><summary>Screenshot</summary>

image
</details> 

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

### CODE REVIEW

File: .github/workflows/deploy-standalone-plugins.yml [hunk 1]
```diff
@@ -12,88 +13,81 @@ on:
...
         with:
           node-version-file: '.nvmrc'
           cache: npm
+
       - name: Install npm dependencies
-        run: npm ci
-      - name: Building standalone plugins
-        if: ${{ matrix.directory != 'plugins' }}
-        run: npm run build-plugins
-      - name: Set version
```
Comment (thelovekesh) [inline .github/workflows/deploy-standalone-plugins.yml:81]: @felixarntz @mukeshpanchal27 Can you please confirm if need to set this version? I think that was required for modules?

File: .github/workflows/deploy-dotorg.yml [hunk 2]
```diff
@@ -7,17 +7,33 @@ on:
...
+        run: npm run build:plugin:performance-lab
+
       - name: WordPress plugin deploy
         uses: 10up/action-wordpress-plugin-deploy@stable
         env:
           SVN_PASSWORD: ${{ secrets.SVN_PASSWORD }}
           SVN_USERNAME: ${{ secrets.SVN_USERNAME }}
+          BUILD_DIR: 'build/performance-lab'
+          ASSETS_DIR: '.wordpress-org'
           SLUG: performance-lab
```
Comment (felixarntz) [inline .github/workflows/deploy-dotorg.yml]: What is the purpose of all these changes? I don't see a good reason for them. The main Performance Lab process has not needed a build process so far, and I don't see why it would need one now.

I understand we need to fix the process for standalone plugins, but for Performance Lab it has worked perfectly well the way it has been so far. The additions here, if anything, will make the process slower as far as I can tell, due to installing NPM dependencies etc.

File: .distignore [hunk 3]
```diff
@@ -0,0 +1,87 @@
+############
```
Comment (felixarntz) [inline .distignore]: Similar to my questions elsewhere: Why `.distignore` instead of `.gitattributes`?

`.gitattributes` has the benefit that it is also processed by GitHub. When you click "Download ZIP", you only get a ZIP of the actual Performance Lab plugin without the dev ops files. With `.distignore` that is no longer the case.

If we replace `.gitattributes` with `.distignore`, we need a compelling reason why. Can you explain your rationale?

File: package.json [hunk 4]
```diff
@@ -12,18 +12,26 @@
...
+    "webpackbar": "^6.0.1"
   },
   "scripts": {
     "changelog": "./bin/plugin/cli.js changelog",
     "since": "./bin/plugin/cli.js since",
     "readme": "./bin/plugin/cli.js readme",
     "translations": "./bin/plugin/cli.js translations",
-    "build-plugins": "./bin/plugin/cli.js build-plugins",
+    "build-plugins": "npm-run-all build:plugin:*",
+    "build:plugin:performance-lab": "webpack --mode production --env plugin=performance-lab",
```
Comment (felixarntz) [inline package.json]: See above: I don't see why we need a build process for Performance Lab. Its process works perfectly fine with only relying on `.gitattributes` for the files and directories to ignore.

File: .github/workflows/deploy-standalone-plugins.yml (hunk 1 above)
Comment (felixarntz) [reply .github/workflows/deploy-standalone-plugins.yml:81]: This is needed to set the correct `VERSION` environment variable for the 10up deployment workflow. You removed this below, but it has to be maintained to make sure we publish the correct version.

File: .github/workflows/deploy-standalone-plugins.yml [hunk 5]
```diff
@@ -12,88 +13,81 @@ on:
...
+      - name: Deploy Standalone Plugin - ${{ matrix.plugin }}
         uses: 10up/action-wordpress-plugin-deploy@stable
         with:
-          dry-run: ${{ matrix.dry-run }}
+          dry-run: ${{ github.event_name == 'workflow_dispatch' && inputs.dry-run || false }}
         env:
           SVN_PASSWORD: ${{ secrets.SVN_PASSWORD }}
           SVN_USERNAME: ${{ secrets.SVN_USERNAME }}
-          SLUG: ${{ matrix.slug }}
-          VERSION: ${{ steps.set_version.outputs.version }}
```
Comment (felixarntz) [inline .github/workflows/deploy-standalone-plugins.yml:97]: See above, this must not be removed.

File: .github/workflows/deploy-standalone-plugins.yml [hunk 6]
```diff
@@ -12,88 +13,81 @@ on:
...
+
 jobs:
-  release:
-    name: Prepare Deployment
+  pre-run:
+    name: Pre-run
     runs-on: ubuntu-latest
     outputs:
-      matrix: ${{ steps.set-matrix.outputs.matrix }}
+      matrix: ${{ steps.set-matrix.outputs.json }}
```
Comment (felixarntz) [inline .github/workflows/deploy-standalone-plugins.yml]: What does this change do?

**Review by felixarntz — CHANGES_REQUESTED** [review]
@thelovekesh For building the standalone plugins, this looks solid so far.

However, I don't think it makes sense to introduce a build process for Performance Lab, this is a solution in search of a problem IMO. And I don't see a good reason to use `.distignore` over `.gitattributes`, please clarify why you're proposing this change.

File: .github/workflows/deploy-standalone-plugins.yml [hunk 7]
```diff
@@ -12,88 +13,65 @@ on:
...
+
       - name: Set matrix
         id: set-matrix
         run: |
+          PLUGINS=$(jq -r '.plugins' plugins.json)
           if ${{ github.event_name == 'workflow_dispatch' }}; then
-            result=$(echo "${{ steps.get-version.outputs.version }}" | awk '/^(\*|[0-9]+(\.[0-9]+){0,2}(-[a-zA-Z0-9.]+)?)$/ {print "Matched"}')
-            if [[ -n "$result" ]]; then
-              # Set the manual input values in JSON format for use in the matrix.
-              echo "matrix={\"include\":[{\"slug\":\"${{ inputs.slug }}\",\"version\":\"${{ steps.get-version.outputs.version }}\",\"directory\":\"${{ steps.get-plugin-directory.outputs.directory }}\",\"dry-run\":\"${{ inputs.dry-run }}\"}]}" >> $GITHUB_OUTPUT
```
Comment (mukeshpanchal27) [inline .github/workflows/deploy-standalone-plugins.yml:48]: Why the manual workflow is removed? https://github.com/WordPress/performance/actions/workflows/deploy-standalone-plugins.yml?query=branch%3Aupdate%2Fplugin-bundling

**Review by mukeshpanchal27 — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy-standalone-plugins.yml (hunk 7 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:48]: @mukeshpanchal27 Still working on the workflow files. Will update you once done.

File: .github/workflows/deploy-standalone-plugins.yml (hunk 1 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:81]: Thanks for confirming. I have re-added this. Now the Webpack generates the build manifest of plugin(s) which can provide version information.

<details><summary>manifest.json</summary>

```json
{
  "performance-lab": "2.9.0",
  "auto-sizes": "1.0.1",
  "dominant-color-images": "1.0.2",
  "speculation-rules": "1.1.0",
  "webp-uploads": "1.0.6"
}
```

</details>

File: .github/workflows/deploy-dotorg.yml (hunk 2 above)
Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: @felixarntz Thanks for pointing this out. I have updated the PR description with the required details on why we need to switch to another build setup.

Now as you mentioned:

>.gitattributes has the benefit that it is also processed by GitHub. When you click "Download ZIP", you only get a ZIP of the actual Performance Lab plugin without the dev ops files. With .distignore that is no longer the case.

In cases where the PL plugin has such files that are not in the git tree and are generated by a build setup, this thing will not be the ideal solution. IMO prod-ready zip should be tagged to plugin release assets which also provide the flexibility to download releases with GitHub APIs.

Now a build step for the main plugin brings a few things to the table:
- A prod-ready zip is tagged on the release assets.
- In case someone needs the PL plugin built locally, it provides the way to do it.
- Since we already created a build system for standalone plugins, the same will also work with the main plugin. So I think it would be better to take the control over distribution creation process of the PL plugin as well from deploy action like we are doing for other plugins.

File: .distignore (hunk 3 above)
Comment (thelovekesh) [reply .distignore]: Please check my above comment and https://github.com/WordPress/performance/issues/893#issuecomment-1947398713

File: .github/workflows/deploy-standalone-plugins.yml (hunk 6 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml]: It simplifies the matrix for the plugins list. Now we just provide the array of plugins in the matrix and all other data is fetched from build manifest.

File: .github/workflows/deploy-dotorg.yml (hunk 2 above)
Comment (felixarntz) [reply .github/workflows/deploy-dotorg.yml]: @thelovekesh 
> In cases where the PL plugin has such files that are not in the git tree and are generated by a build setup, this thing will not be the ideal solution.

Can you provide a concrete example where that is the case? I don't see any files being generated by the build setup for Performance Lab.

> IMO prod-ready zip should be tagged to plugin release assets which also provide the flexibility to download releases with GitHub APIs.

That is a good point, but it doesn't require us to have a build step. This can be accomplished with the 10up deploy action alone, see https://github.com/10up/action-wordpress-plugin-deploy?tab=readme-ov-file#inputs.

I think the reasons you're pointing out are valid in theory, but not in practice. As far as I can tell, the Performance Lab plugin doesn't require to be built, it works just from the source - unless I'm missing something. Unless there is a concrete usage that requires the build process, I think we should avoid it, as it introduces a process for something that doesn't need it, which makes things slower and is one more thing to maintain.

File: .distignore (hunk 3 above)
Comment (felixarntz) [reply .distignore]: Replied to your reply there :)

File: .github/workflows/deploy-dotorg.yml (hunk 2 above)
Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: @felixarntz I think there are two possible scenarios here:
- PL plugin which doesn't have any build setup as of now, so works well with the current `.gitattributes` based setup.
- Standalone plugins, where:
    - ILO and Web Worker Offloading have a build step that includes vendor library.
    - We need to generate builds for these which should have the vendor files(which are git ignored) and should not have config/WP Assets files.

Now we have a webpack-based build step(by keeping standalone plugins in mind) that will build plugins and ignore files that are not required but include files that are git ignored(which doesn't seem possible with `.gitattributes`), hence we switched to `.distignore`.

> Unless there is a concrete usage that requires the build process, I think we should avoid it, as it introduces a process for something that doesn't need it, which makes things slower and is one more thing to maintain.

Now the same build setup can be used to generate for build PL plugin as well at no extra cost, I don't think it will require any maintenance.

Comment (felixarntz) [reply .github/workflows/deploy-dotorg.yml]: @thelovekesh Based on your reply, there's no reason to use a build process for Performance Lab. I completely agree the other standalone plugins need it, but that doesn't mean we need to use it for PL as well. Even if the maintenance cost is little, there is no benefit to using it for Performance Lab. So let's use it only for the standalone plugins.

> Now we have a webpack-based build step(by keeping standalone plugins in mind) that will build plugins and ignore files that are not required but include files that are git ignored(which doesn't seem possible with `.gitattributes`), hence we switched to `.distignore`.

I don't understand this. `.gitattributes` is used by GitHub e.g. to ignore files when users download a ZIP file, and it's also used by the 10up deploy script. I don't think it interferes with a custom Webpack build process. Of course 3rd party libraries like `web-vitals` need to be in `.gitignore`, which means they don't even need to be listed in `.gitattributes` because they won't be part of the repo. For the build process of the standalone plugins, that means they have to run `npm install` before building the standalone plugins, and those aren't subject to `.gitattributes` anyway - only the main PL plugin is.

In other words: I think as long as we continue to separate the processes, one for deploying PL, another one for building and deploying standalone plugins, there should not be an issue. If I'm missing something, can you please provide more detail on which problem you're encountering because of `.gitattributes`?

Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: > In other words: I think as long as we continue to separate the processes, one for deploying PL, and another one for building and deploying standalone plugins, there should not be an issue. If I'm missing something, can you please provide more detail on which problem you're encountering because of .gitattributes?

Okay, I see. In essence, my goal was to make this procedure more efficient for every plugin. I believe that if we leave `.gitattributes` alone, I will have to explicitly specify which patterns to ignore for the standalone plugin build rather than obtaining them from a global `.distignore`.

I don't mind if the release workflow for PL and standalone plugins is kept separate, but I still think that the build setup should retain `.distignore` since the 10up deploy action [supports](https://github.com/10up/action-wordpress-plugin-deploy?tab=readme-ov-file#excluding-files-from-deployment) it (if you want to handle the build from the 10up action) or with plugin build setup if you agree to keep `.distignore`. 

Attaching the build zip to the release section is always an option if we are only keeping the `.gitattributes` because it offers processed ZIP. In any case, once we have build files for the PL plugin, we will need to delete it. Who knows, maybe we'll use WordPress components for https://github.com/WordPress/performance/issues/1032. However, as you pointed out, that's only a theory 😎

Comment (felixarntz) [reply .github/workflows/deploy-dotorg.yml]: > I believe that if we leave `.gitattributes` alone, I will have to explicitly specify which patterns to ignore for the standalone plugin build rather than obtaining them from a global `.distignore`.

This is the part I do not understand. How does this work when using `.distignore` but not when using `.gitattributes`? As far as my understanding goes, neither of these two files are used by the 10up deploy action when we specify a custom `BUILD_DIR`, as you pointed out before. So what's the difference between these two files for the build process you're adding here? Where in the build process is `.distignore` used? Maybe I'm missing something, so can you please clarify?

Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: Aah okay. I am using `.distignore` to pass patterns to be ignored in the `copy-webpack-plugin`.

https://github.com/WordPress/performance/blob/9f38e400ecdac6ddf3abc505adf0d44e11db273b/webpack.config.js#L82

Comment (felixarntz) [reply .github/workflows/deploy-dotorg.yml]: I see! Thanks for clarifying. From that perspective, I agree using `.distignore` makes sense.

However, we would still lose out on the current default GitHub behavior from using `.gitattributes` here.

How about the following:
* We keep `.gitattributes` as is for now.
* We use `.distignore` too, but _only_ for the standalone plugins. This means that list can be a lot shorter (probably only needs to cover 3-5 file/folder names and it doesn't have to be a "copy" of `.gitattributes`, which avoids a maintenance nightmare. I think you'd only need to put `.wordpress-org`, `phpcs.xml.dist`, and `.distignore` in there.

I'm not sure this is necessarily a great end solution, but I think it would be good enough for now and satisfy all our use-cases. As long as Performance Lab doesn't need a build process, it doesn't need to use `.distignore` either, so we could limit `.distignore` to only the few files/folders used for the standalone plugins. We should add a comment on top of `.distignore` to clarify this is only used for the standalone plugin build process.

WDYT?

Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: Yes @felixarntz, it makes sense to me. I believe we should keep the `.gitattributes` file as it is also processed by https://github.com/github-linguist/linguist (and the AMP plugin uses it like https://github.com/ampproject/amp-wp/blob/develop/.gitattributes). This is after researching the usage of `.gitattributes` in a GitHub repository in greater detail.

The only drawback is that the 10up deploy action will prioritize `.distignore` if we retain both `.gitattributes` and `.distignore`([ref](https://github.com/10up/action-wordpress-plugin-deploy/blob/abb939a0d0bfd01063e8d1933833209201557381/deploy.sh#L94)). 

Since Webpack is now using `.distignore`, I would suggest moving it to `bin/webpack`. As a result, the standalone plugin will continue to use `.distignore`, and the PL plugin will continue to use `.gitattributes`. 

We can easily transfer it to the plugin root if we ever decide to use `.distignore` for PL plugins as well.

Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: Also, I don't see any issues if we use `.distignore` with the 10up deploy action as well. Then it will be like:
- Use `.distignore` for creating distribution builds.
- Use `.gitattributes` for GitHub related chores and ZIP processing.

Comment (felixarntz) [reply .github/workflows/deploy-dotorg.yml]: @thelovekesh 
> Since Webpack is now using `.distignore`, I would suggest moving it to `bin/webpack`. As a result, the standalone plugin will continue to use `.distignore`, and the PL plugin will continue to use `.gitattributes`. 

This sounds great to me.

> Also, I don't see any issues if we use `.distignore` with the 10up deploy action as well. Then it will be like:
> - Use `.distignore` for creating distribution builds.
> - Use `.gitattributes` for GitHub related chores and ZIP processing.

While we could do that, this would mean we would still need to basically maintain the large file list in two places. So I would prefer to go with your prior suggestion and use `.gitattributes` in the root (for both GitHub and Performance Lab build/deployment) and `.distignore` in `bin/webpack` (for the standalone plugins build/deployment). That way we can keep `.distignore` light and only focused on the few files that need to be ignored for the standalone plugin folders.

Comment (thelovekesh) [reply .github/workflows/deploy-dotorg.yml]: @felixarntz It looks like the ignore file is not working as expected when moving it to `bin/webpack` directory since it's adding cwd path to patterns. Hence I have directally added the ingore paths in Webpack config for now since they are very few. Please let me know if it looks good to you.

https://github.com/WordPress/performance/blob/41b45ca5826671fe5cb7599704a688310336d0d3/webpack.config.js#L68-L71

Comment (felixarntz) [reply .github/workflows/deploy-dotorg.yml]: @thelovekesh Thanks, that works for me! Given it's so few files, it should be okay.

File: package.json [hunk 8]
```diff
@@ -12,18 +12,26 @@
...
+    "webpackbar": "^6.0.1"
   },
   "scripts": {
     "changelog": "./bin/plugin/cli.js changelog",
     "since": "./bin/plugin/cli.js since",
     "readme": "./bin/plugin/cli.js readme",
     "translations": "./bin/plugin/cli.js translations",
-    "build-plugins": "./bin/plugin/cli.js build-plugins",
+    "build-plugins": "npm-run-all 'build:plugin:!(performance-lab)'",
+    "build:plugin:performance-lab": "rm -rf build/performance-lab && mkdir -p build/performance-lab && git archive HEAD | tar -x -C build/performance-lab",
```
Comment (thelovekesh) [inline package.json:30]: @felixarntz Added this command, in case we ever need to confirm what's being added in the build by deploy action.

File: .github/workflows/deploy-standalone-plugins.yml [hunk 9]
```diff
@@ -1,99 +1,154 @@
...
     steps:
-      - name: Checkout repository
-        uses: actions/checkout@v3
-      - name: Setup Node.js (.nvmrc)
-        uses: actions/setup-node@v3
-        with:
-          node-version-file: '.nvmrc'
-          cache: npm
-      - name: Install npm dependencies
-        run: npm ci
```
Comment (felixarntz) [inline .github/workflows/deploy-standalone-plugins.yml:30]: You're moving these steps to `deploy` further below, which runs for every plugin to deploy. Maybe that's the only way, but would it be possible to instead leave these actions here?

The reason is that checking out the repository and installing dependencies will be the same for every plugin. So it would be more efficient to do it once in the beginning, regardless of how many or which plugins are being deployed afterwards.

File: .github/workflows/deploy-standalone-plugins.yml [hunk 10]
```diff
@@ -1,99 +1,154 @@
...
+      - name: Set plugin version
+        working-directory: ./build
         run: |
-          if [ -z "${{ matrix.version }}" ]; then
-            echo "version=$(node ./bin/plugin/cli.js get-plugin-version --slug=${{ matrix.slug }})" >> $GITHUB_OUTPUT
-          else
-            echo "version=${{ matrix.version }})" >> $GITHUB_OUTPUT
-          fi
-      - name: Deploy Standalone Plugin - ${{ matrix.slug }}
+          echo "PLUGIN_VERSION=$(jq -r '."${{ matrix.plugin }}"' manifest.json)" >> $GITHUB_ENV
```
Comment (felixarntz) [inline .github/workflows/deploy-standalone-plugins.yml:78]: Curious: Why use `$GITHUB_ENV` instead of `$GITHUB_OUTPUT` like before?

**Review by felixarntz — COMMENTED** [review]
@thelovekesh This looks quite good to me now in terms of the overall approach. A few small comments/questions below.

I'm not very familiar with GitHub workflows, so it would be great to get @swissspidy's feedback on those changes in particular.

File: .github/workflows/deploy-standalone-plugins.yml (hunk 7 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:48]: @mukeshpanchal27 I have updated the workflow files. Please feel free to re-review them.

Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:48]: And JFI, we are now using build manifest instead of using a CLI script to determine the plugin versions.

File: .github/workflows/deploy-standalone-plugins.yml (hunk 10 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:78]: I think it's best to use `$GITHUB_OUTPUT` when you want to share the info across the jobs. For the same job, an env variable is suffice.

File: .github/workflows/deploy-standalone-plugins.yml (hunk 9 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:30]: @felixarntz These steps are here just to install NPM deps so that we can run a few CLI commands to get the plugin version and path to the matrix.

In the previous setup, we were running `npm ci` in two places:
- release job
- deployment job(which is also a matrix)

In the current setup, we were running `npm ci`  only once:
- deploy job (which is a matrix)

File: package.json (hunk 8 above)
Comment (thelovekesh) [reply package.json:30]: @felixarntz I believe that we should use the above command to verify distribution files(maybe just before a release?). I can see that the `build-cs` directory is being committed in the svn at https://plugins.trac.wordpress.org/browser/performance-lab/trunk for the most current release 😞. 

Also, it would be good if we log an issue to remove such redundant files/dirs from all our plugins in the next release.

File: .github/workflows/deploy-standalone-plugins.yml [hunk 11]
```diff
@@ -1,99 +1,154 @@
...
+          cd ./build/${{ matrix.plugin }}
+          zip -r ../dist/${{ matrix.plugin }}.zip .
+
+      - name: Generate checksum
+        working-directory: ./build/dist
+        run: |
+          mkdir -p $RUNNER_TEMP/plugin-checksums
+          find . -type f -print0 | sort -z | xargs -r0 shasum -a 256 -b | sed 's# \*\./# *#' > $RUNNER_TEMP/plugin-checksums/checksums.txt
+          shasum -a 256 -U -c $RUNNER_TEMP/plugin-checksums/checksums.txt
+          cat $RUNNER_TEMP/plugin-checksums/checksums.txt | while read sum file; do echo "$sum $file" > ${file#\*}.sha256; done
```
Comment (swissspidy) [inline .github/workflows/deploy-standalone-plugins.yml]: Do we really need to create checksums? A bit overkill IMO. I doubt anyone would ever look at these.

If someone wants checksum verification, WordPress.org already provides that information anyway through a nice API.

File: .github/workflows/deploy-standalone-plugins.yml [hunk 12]
```diff
@@ -1,99 +1,154 @@
...
+
+      - name: Upload artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ matrix.plugin }}
+          path: ./build/dist
+
+      - name: Start deployment
+        uses: bobheadxi/deployments@v1
```
Comment (swissspidy) [inline .github/workflows/deploy-standalone-plugins.yml:126]: Interesting action! 👍 👀

**Review by swissspidy — APPROVED** [review]
Overall looks good 👍  Looking forward to seeing this in action.

**Review by mukeshpanchal27 — COMMENTED** [review]
@thelovekesh, thank you for the updates.

It seems that `npm run test-plugins` is broken. Have you checked the automated and manual plugin release workflow with `dry-run` set to true? This way, we can thoroughly test all steps of the workflow. Could you please ensure the following:

- Added a pull_request for checking the automated workflow.
- Set both auto and manual workflow DRY Run to true for security.
- Updated SVN credentials for double security.

**Comment by thelovekesh** [comment]
Thanks @mukeshpanchal27. It seems like breaking changes are unrelated to this PR. I can see them on https://github.com/WordPress/performance/pull/1028 as well - https://github.com/WordPress/performance/actions/runs/8182298569/job/22373438106

>  Have you checked the automated and manual plugin release workflow with dry-run set to true?

Yes, I have tested them on my fork. 
- Standalone plugins release logs - https://github.com/thelovekesh/performance/actions/workflows/deploy-standalone-plugins.yml
- PL plugin release logs - https://github.com/thelovekesh/performance/actions/workflows/deploy-dotorg.yml
- Release assets linking - https://github.com/thelovekesh/performance/releases/tag/test-release
- Deployment status - https://github.com/thelovekesh/performance/deployments/

File: .github/workflows/deploy-standalone-plugins.yml (hunk 11 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml]: It's not a "must use," but providing a means to verify the integrity of your distributed assets is never bad.

File: .gitattributes [hunk 13]
```diff
@@ -10,6 +10,7 @@
 /tests export-ignore
 /plugin-tests export-ignore
 /plugins export-ignore
+/build-cs export-ignore
```
Comment (felixarntz) [inline .gitattributes:13]: Great catch!

File: .github/workflows/deploy-standalone-plugins.yml [hunk 14]
```diff
@@ -1,99 +1,154 @@
...
+        with:
+          name: ${{ matrix.plugin }}
+          path: ./build/dist
+
+      - name: Upload release assets
+        uses: softprops/action-gh-release@v1
+        with:
+          files: |
+            ./build/dist/${{ matrix.plugin }}.zip
+            ./build/dist/${{ matrix.plugin }}.zip.sha256
```
Comment (felixarntz) [inline .github/workflows/deploy-standalone-plugins.yml]: As far as I can tell, this will add the release assets to _any_ GitHub release, correct? Does that mean when we publish a new Performance Lab release and all relevant standalone plugins are updated as part of that, the Performance Lab release on GitHub will have all those standalone plugins' ZIP files attached?

I'm not saying this is necessarily a bad idea, I just want to make sure we have thought this through. In the future, we may consider having individual GitHub releases for the standalone plugins, so in such a scenario we probably wouldn't want to do this, but rather attach the ZIP file only to the corresponding release of _that_ plugin.

**Review by felixarntz — COMMENTED** [review]
@thelovekesh @swissspidy How does this PR determine which plugins to deploy an update for? Here's what I mean:
* When a new Performance Lab release is triggered for instance, not all standalone plugins will need to be updated.
* A plugin should only be updated if its new version is greater than the latest version currently available on wordpress.org.

I'm not certain the workflow currently accounts for that. I know the 10up deploy action handles this internally already, but I don't see anything in the other steps of our workflow that take that into account. Can you double check? 🤔 

Also, another somewhat related comment on which ZIP files get attached to which GitHub release. FWIW, if that complicates getting this merged, we could always leave that out and add separately, also to solve it holistically for Performance Lab as well. We need to rethink our release strategy anyway.

File: .github/workflows/deploy-standalone-plugins.yml (hunk 14 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml]: > As far as I can tell, this will add the release assets to any GitHub release, correct? Does that mean when we publish a new Performance Lab release and all relevant standalone plugins are updated as part of that, the Performance Lab release on GitHub will have all those standalone plugins' ZIP files attached?

That's correct. 

> I'm not saying this is necessarily a bad idea, I just want to make sure we have thought this through. In the future, we may consider having individual GitHub releases for the standalone plugins, so in such a scenario we probably wouldn't want to do this, but rather attach the ZIP file only to the corresponding release of that plugin.

@felixarntz As long as we are releasing all plugins on release, IMO we can keep it. As mentioned in https://wordpress.slack.com/archives/C02KGN5K076/p1710262256960409, once we rethink on how we streamline the release this can be taken care of.

Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml]: Also, it would be better if we open a separate ticket for this and track all the discussions there.

**Comment by thelovekesh** [comment]
> I'm not certain the workflow currently accounts for that. I know the 10up deploy action handles this internally already, but I don't see anything in the other steps of our workflow that take that into account. Can you double check? 🤔

@felixarntz At this point, no it doesn't take that into account and it's that way only. We need to re-consider on how we release our plugins and IMO that should be a part of different issue/PR.

**Comment by felixarntz** [comment]
@thelovekesh 
> > I'm not certain the workflow currently accounts for that. I know the 10up deploy action handles this internally already, but I don't see anything in the other steps of our workflow that take that into account. Can you double check? 🤔
> 
> @felixarntz At this point, no it doesn't take that into account and it's that way only. We need to re-consider on how we release our plugins and IMO that should be a part of different issue/PR.

I agree with you regarding the attachment of ZIP files to the release that it can be solved separately. However, the part about detecting whether a plugin even needs to be updated has to be handled here, otherwise it breaks existing behavior. For example:
* Let's say "WebP Uploads" is currently live on wordpress.org in version 1.0.5.
* In this repository, it currently has version 1.0.5 in its plugin main file and `readme.txt`.
* Publishing a new release on GitHub must not attach a ZIP file for WebP Uploads to that release, because 1.0.5 was already published before.

In other words, this would be incorrect since the 1.0.5 ZIP generated wouldn't _actually_ be the 1.0.5 release. That's why this needs to be fixed here to prevent this bug.

**Comment by thelovekesh** [comment]
sure @felixarntz. I have updated the workflow only to deploy a plugin if it's not already deployed on the WPOrg.

File: .github/workflows/deploy-standalone-plugins.yml (hunk 14 above)
Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml]: Addressed in https://github.com/WordPress/performance/pull/1033/commits/0d5451d8bcd9c4619a1ccce54bf7e6e981b5987c

**Review by felixarntz — APPROVED** [review]
@thelovekesh Looks great to me!

Before merge, @swissspidy could you take another brief look at the new additions since you had previously approved?

File: .github/workflows/deploy-standalone-plugins.yml [hunk 15]
```diff
@@ -1,99 +1,201 @@
...
+          name: ${{ matrix.plugin }}
+          path: ./build/dist
+
+      - name: Upload release assets
+        if: steps.artifact-existence.outputs.exists == 'true'
+        uses: softprops/action-gh-release@v1
+        with:
+          files: |
+            ./build/dist/${{ matrix.plugin }}.zip
+            ./build/dist/${{ matrix.plugin }}.zip.sha256
```
Comment (westonruter) [inline .github/workflows/deploy-standalone-plugins.yml:201]: So this will attach all standalone plugin ZIPs to the Performance Lab plugin release? If so, it seems like a missing piece in general for the standalone plugin story is how to trigger a release of a standalone plugin without having to do a release of Performance Lab as well. I guess that can still be done by [manually invoking the workflow](https://github.com/WordPress/performance/actions/workflows/deploy-standalone-plugins.yml), but then there would be no record of the standalone plugin release among the GitHub releases.

Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:201]: Yes, that's right. For now, there are two scenarios:
- Release trigger.
- On workflow dispatch.

In the second case, there will be no release record as mentioned by @westonruter. In the last Performance chat, we discussed opening a new issue which I have just opened https://github.com/WordPress/performance/issues/1061 to discuss it.

Comment (felixarntz) [reply .github/workflows/deploy-standalone-plugins.yml:201]: @westonruter Correct, at the moment the manual workflow is the way to do it, if we need to release something apart from a PL release.

A discussion on how to better address that is long overdue and has come up in a few places before. Let's tackle that in a separate issue. Potentially we can just create GitHub releases for the individual plugins and have the workflows adjust based on e.g. the release name or something like that what gets deployed.

Comment (thelovekesh) [reply .github/workflows/deploy-standalone-plugins.yml:201]: > So this will attach all standalone plugin ZIPs to the Performance Lab plugin release?

Only for such plugins that are not already released on WPOrg. See https://github.com/WordPress/performance/pull/1033/commits/0d5451d8bcd9c4619a1ccce54bf7e6e981b5987c

File: bin/webpack/utils.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+ * Determine the plugin version from the readme.txt file.
+ *
+ * @param {string} pluginPath The path to the plugin.
+ *
+ * @return {string|false} The plugin version or false if not found.
+ */
+const getPluginVersion = ( pluginPath ) => {
+	const readmePath = path.resolve( pluginPath, 'readme.txt' );
+
+	const fileContent = fs.readFileSync( readmePath, 'utf-8' );
```
Comment (mukeshpanchal27) [inline bin/webpack/utils.js:36]: Not blocker: Better to use try catch for error handling.

```
let fileContent = '';
try {
	fileContent = fs.readFileSync( readmePath, 'utf-8' );
} catch ( error ) {
	log(
		formats.error(
			`Error reading the file "${ readmePath }": "${ error }"`
		)
	);
}
```

Comment (thelovekesh) [reply bin/webpack/utils.js:36]: There's no reason to handle the error here, in my opinion. We must end the operation here if it is unable to read or resolve the `readme.txt`. We don't need to add our error statements as NodeJS errors are already verbose.

**Review by swissspidy — APPROVED** [review]
Still looks good to me now. I'd say let's just get this in and give it a try, we can always refine later if needed.

**Review by mukeshpanchal27 — APPROVED** [review]
